File: PROGRAMMING FOUNDATION - PYTHON/LIBRARY_MANAGEMENT_SYSTEM/Member.py

class Person:
    def __init__(self, name):
        self._name = name

class Member(Person):
    def __init__(self, name, member_id):
        super().__init__(name)
        self._name=name
        self._member_id = member_id
        self._borrowed_books = [] 
        print(f"Member Name-{self._name} of ID-{self._member_id} added to member list\n")
  
    def borrow_book(self, book):
        if book.get_isbn() in self._borrowed_books:
            raise Exception("Member already borrowed this book.")

        self._borrowed_books.append(book.get_isbn())
        print(f"Book '{book.get_title()}({book.get_isbn()})'  issued to '{self.get_name()}({self.get_member_id()})'.")

    def return_book(self, book):
        if book.get_isbn() not in self._borrowed_books:
            raise Exception("Book not in member's borrowed list.")

        self._borrowed_books.remove(book.get_isbn())
        print(f"Book '{book.get_title()}({book.get_isbn()})'  returned by '{self.get_name()}({self.get_member_id()})'\n")

    def get_member_id(self):
        return self._member_id
    def get_name(self):
        return self._name

    def __str__(self):

        return f"Member: {self._name} | ID: {self._member_id}"

File: PROGRAMMING FOUNDATION - JAVA/model/Member.java
package model;

import java.time.LocalDate;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;

class Member {
    private String memberId;
    private String name;
    private ArrayList<Book> issuedBooks;
    private ArrayList<LocalDate> issuedDates;

    public Member( String memberId,String name) {
        this.memberId = memberId;
        this.name = name;
        issuedBooks = new ArrayList<>();
        issuedDates = new ArrayList<>();
    }

    public String getMemberId() {
        return memberId;
    }

    public void issueBook(Book b) {
        if (!b.isAvailable()) {
            issuedBooks.add(b);
            issuedDates.add(LocalDate.now());
            b.markUnavailable();
            System.out.println(name + " issued: " + b.getBookTitle());
        } else {
            System.out.println(b.getBookTitle() + " is not available right now.");
        }
    }

    public void returnBook(Book b) {
        int idx = issuedBooks.indexOf(b);
        if (idx >= 0) {
            LocalDate takenDate = issuedDates.get(idx);
            long diff = ChronoUnit.DAYS.between(takenDate, LocalDate.now());
            double penalty = diff > 15 ? (diff - 15) * 2 : 0;

            issuedBooks.remove(idx);
            issuedDates.remove(idx);
            b.markAvailable();

            System.out.println("'"+b.getBookTitle() + "' returned by : " + name + " | Fine: ₹" + penalty);
        } else {
            System.out.println(name + " did not take this book.");
        }
    }

    public void displayUser() {
        System.out.println("User: " + name + " | MemberID: " + memberId);
        System.out.println("Issued Books:");
        for (Book b : issuedBooks)
            System.out.println("• " + b.getBookTitle());
    }
}
